core/src/main/java/org/apache/druid/timeline/partition/OvershadowableManager.java

```diff
@@ -0,0 +1,780 @@
+      } else {
+        atomicUpdateGroup = searchForStateOf(chunk, State.VISIBLE);
+
+        if (atomicUpdateGroup != null) {
+          // A new chunk of the same major version and partitionId can be added in segment handoff
+          // from stream ingestion tasks to historicals
+          final PartitionChunk<T> existing = atomicUpdateGroup.replaceChunkWith(chunk);
+          if (existing == null) {
+            throw new ISE(
+                "Can't add a new partitionChunk[%s] to a visible atomicUpdateGroup[%s]",
+                chunk,
+                atomicUpdateGroup
+            );
+          } else if (!chunk.equals(existing)) {
+            throw new ISE(
+                "WTH? a new partitionChunk[%s] has the same partitionId but different from existing chunk[%s]",
+                chunk,
+                existing
+            );
+          }
+        } else {
+          final AtomicUpdateGroup<T> newAtomicUpdateGroup = new AtomicUpdateGroup<>(chunk);
+
+          // Decide the initial state of the new atomicUpdateGroup
+          final boolean overshadowed = visibleGroup
+              .values()
+              .stream()
+              .flatMap(map -> map.values().stream())
+              .anyMatch(group -> group.isOvershadow(newAtomicUpdateGroup));
+
+          if (overshadowed) {
+            addTo(newAtomicUpdateGroup, State.OVERSHADOWED);
+          } else {
+            addTo(newAtomicUpdateGroup, State.STANDBY);
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Handles of removal of an empty atomicUpdateGroup from a state.
+   */
+  private void handleRemove(
+      AtomicUpdateGroup<T> augOfRemovedChunk,
+      RootPartitionRange rangeOfAug,
+      short minorVersion,
+      State stateOfRemovedAug
+  )
+  {
+    if (stateOfRemovedAug == State.STANDBY) {
+      // If an atomicUpdateGroup is overshadowed by another standby atomicUpdateGroup, there must be another visible
+      // atomicUpdateGroup which also overshadows the same atomicUpdateGroup.
+      // As a result, the state of overshadowed atomicUpdateGroup shouldn't be changed and we do nothing here.
+
+    } else if (stateOfRemovedAug == State.VISIBLE) {
+      // All segments in the visible atomicUpdateGroup which overshadows this atomicUpdateGroup is removed.
+      // Fall back if there is a fully available overshadowed atomicUpdateGroup
+
+      final List<AtomicUpdateGroup<T>> latestFullAugs = findLatestFullyAvailableOvershadowedAtomicUpdateGroup(
+          rangeOfAug,
+          minorVersion
+      );
+
+      if (!latestFullAugs.isEmpty()) {
+        // Move the atomicUpdateGroup to standby
+        // and move the fully available overshadowed atomicUpdateGroup to visible
+        if (!augOfRemovedChunk.isEmpty()) {
+          transitPartitionChunkState(augOfRemovedChunk, State.VISIBLE, State.STANDBY);
+        }
+        latestFullAugs.forEach(group -> transitPartitionChunkState(group, State.OVERSHADOWED, State.VISIBLE));
+      }
```

Comment:
Suggest adding a comment here for the isEmpty == true case, saying something like "If there is no fully available fallback group, then the existing VISIBLE group remains VISIBLE"

Reply:
Added.